Separate SourceState/JobState from workUnits

zliu41 opened the pull request:
Summary:
- Deprecated all constructors of `WorkUnit`, and added some factory methods, which do not add job states to workunits. 
The reason to use factory methods is to avoid having too many constructors for `WorkUnit`, some deprecated and some not deprecated.
- Deprecated the constructor of `Extract` that takes `SourceState`, added a new constructor without `SourceState`.
- In `MRJobLauncher.prepareHadoopJob()`, serialize `jobState` into a file. Then in `TaskRunner.setup()`, read `jobState` from the file and add it to each `workUnit`.
- In `LocalJobLauncher.runWorkUnits()`, add `jobState` to each `workUnit`.
- In `MRJobLauncher.runWorkUnits()`, combine `jobState` with each `workUnit` into a `fatWorkUnit`, since properties from both `jobState` and `workUnit` are needed in order to invoke `JobLauncherUtils.cleanStagingData()`.
- Modified `KafkaSource` to use the new paradigm.

Tested with Kafka adapter. Existing source implementations still work.

zliu41 commented:
@liyinan926 @chavdar please review.

liyinan926 commented:
Seems all `MRJobLauncherTest` tests are failing. @zliu41 can you first resolve those?

zliu41 commented:
@liyinan926 updated. Since `WriterUtils.getDefaultWriterFilePath` requires the input `State` object to be either `WorkUnit` or `WorkUnitState`, I needed to change `State combinedState` to `WorkUnit fatWorkUnit`.

liyinan926 commented on gobblin-api/src/main/java/gobblin/configuration/SourceState.java:
```diff
@@ -74,7 +74,7 @@ public SourceState(State properties, List<WorkUnitState> previousTaskStates) {
    */
   public List<WorkUnitState> getPreviousWorkUnitStates() {
-    return ImmutableList.<WorkUnitState>builder().addAll(this.previousTaskStates).build();
+    return ImmutableList.<WorkUnitState> builder().addAll(this.previousTaskStates).build();
```
Empty space introduced.

liyinan926 commented on gobblin-api/src/main/java/gobblin/configuration/SourceState.java:
```diff
@@ -104,14 +104,17 @@ public synchronized Extract createExtract(Extract.TableType type, String namespa
    * @return a new {@link WorkUnit} instance
+   *
+   * @deprecated Properties in SourceState should not added to a WorkUnit. Having each WorkUnit contain a copy of
+   * SourceState is a waste of memory. Use WorkUnit.create(Extract extract).
```
Use `Use {@link WorkUnit#create(Extract)}`.

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/workunit/Extract.java:
```diff
@@ -62,7 +62,10 @@
    * @param type {@link TableType}
    * @param table table name
+   *
+   * @deprecated Extract does not use any property in SourceState. Use Extract(TableType, namespace, table).
```
Ditto.

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/workunit/WorkUnit.java:
```diff
@@ -76,7 +83,11 @@ public WorkUnit(SourceState state, Extract extract) {
    * @param watermarkInterval a {@link WatermarkInterval} which defines the range of data this {@link WorkUnit} will process.
+   *
+   * @deprecated Properties in SourceState should not be added to a WorkUnit. Having each WorkUnit contain a copy of
+   * SourceState is a waste of memory. Use WorkUnit.create(Extract extract, WatermarkInterval watermarkInterval).
```
Ditto.

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/workunit/WorkUnit.java:
```diff
@@ -56,7 +59,11 @@ public WorkUnit() {
    * @param extract an {@link Extract}
+   *
+   * @deprecated Properties in SourceState should not be added to a WorkUnit. Having each WorkUnit contain a copy of
+   * SourceState is a waste of memory. Use WorkUnit.create(Extract extract).
```
Ditto.

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/workunit/WorkUnit.java:
```diff
@@ -95,13 +106,52 @@ public WorkUnit(SourceState state, Extract extract, WatermarkInterval watermarkI
+  /**
+   * Facotry method.
+   *
+   * @return A WorkUnit with the given Extract.
```
Please use `{@link ...}` for class references.

liyinan926 replied:
Incomplete Javadoc.

liyinan926 commented on gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java:
```diff
@@ -557,6 +582,26 @@ protected void setup(Context context) {
 
+    private JobState readJobState(Context context) throws IOException {
+      JobState jobState = new JobState();
+      if (context.getConfiguration().get(ConfigurationKeys.JOB_STATE_FILE_KEY) == null) {
```
You can use a `Preconditions.checkArgument()` here.

liyinan926 commented on gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java:
```diff
@@ -557,6 +582,26 @@ protected void setup(Context context) {
     }
 
+    private JobState readJobState(Context context) throws IOException {
+      JobState jobState = new JobState();
```
You can move this to after the precondition check.

liyinan926 replied:
You can actually change the member variable to `private final JobState jobState = new JobState()` and then just deserialize into `this.jobState`.

liyinan926 commented on gobblin-api/src/main/java/gobblin/configuration/ConfigurationKeys.java:
```diff
@@ -121,6 +121,7 @@
   public static final String JOB_TRACKING_URL_KEY = "job.tracking.url";
   public static final String FORK_STATE_KEY = "fork.state";
   public static final String METRIC_CONTEXT_NAME_KEY = "metrics.context.name";
+  public static final String JOB_STATE_FILE_KEY = "job.state.file";
```
`job.state.file.path` is a better name.

liyinan926 commented on gobblin-api/src/main/java/gobblin/source/workunit/WorkUnit.java:
```diff
@@ -95,13 +106,52 @@ public WorkUnit(SourceState state, Extract extract, WatermarkInterval watermarkI
+
+  /**
+   * Facotry method.
+   *
```
Ditto. Incomplete Javadoc.

liyinan926 commented:
There's still one potential issue with this approach. If someone sets anything in the `workUnitstate` that is supposed to overwrite the static config in the method `Source.getExtractor(WorkUnitState workUnitstate)`, the changes will be lost when `JobState` gets added back.

zliu41 commented:
@liyinan926 updated.
- Added method `State.addAllIfNotExist()`, so that when combining `jobState` and `workUnit`, `jobState` doesn't override `workUnit`.
- Added method `Extract.getUniqueInstance()` which replaces the deprecated method `SourceState.createExtract()`.

sahilTakiar commented:
I think this may also require changes to `AbstractSource`. The method `getPreviousWorkUnitStatesForRetry` contains some code to use a config `ConfigurationKeys.OVERWRITE_CONFIGS_IN_STATESTORE`. If this config is set to true then all properties in the specified `SourceState` object are added to each `WorkUnitState`. I'm guessing this is something we would want to change too.

zliu41 commented:
Updated.

liyinan926 commented:
LGTM.

chavdar commented on gobblin-api/src/main/java/gobblin/source/workunit/Extract.java:
```diff
@@ -81,14 +88,47 @@ public Extract(SourceState state, TableType type, String namespace, String table
+   * @param table table name
+   * @return a unique {@link Extract} instance
+   */
+  public static Extract getUniqueInstance(TableType type, String namespace, String table) {
```
This looks like a copy/paste of gobblin.configuration.SourceState.createExtract(TableType, String, String) which is not good. We need to remove this function and that static CREATED_INSTANCES object. The extract creation should be owned by the Source.

1) Please move the CREATED_INSTANCES map and this method into a new interface/class ExtractFactory/ExtractFactoryImpl or something along those lines. 

2) Have AbstractSource create/own an instance of ExtractFactoryImpl.

3) Have the KafkaSource use the factory provided by AbstractSource

chavdar commented on gobblin-api/src/main/java/gobblin/configuration/SourceState.java:
```diff
@@ -89,10 +89,13 @@ public SourceState(State properties, List<WorkUnitState> previousTaskStates) {
    * @param table name of the table this extract belongs to
    * @return a new unique {@link Extract} instance
+   *
+   * @Deprecated Use {@link Extract#getUniqueInstance(Extract.TableType, String, String)}
```
This should be deprecated in favor of the AbstractSource ExtractFactoryImpl.

chavdar replied:
For now, you can also make EXTRACT_SET and ExtractFactoryImpl and reuse the implementation in createExtract to avoid code duplication.

zliu41 commented:
@chavdar updated.

chavdar commented on gobblin-api/src/main/java/gobblin/source/workunit/ExtractFactory.java:
```diff
@@ -0,0 +1,49 @@
+   * @param table table name
+   * @return a unique {@link Extract} instance
+   */
+  public Extract getUniqueExtract(TableType type, String namespace, String table) {
```
This needs to be synchronized and the createdInstances need not. There are race conditions in the current implementation.

chavdar commented on gobblin-api/src/main/java/gobblin/source/workunit/ExtractFactory.java:
```diff
@@ -0,0 +1,49 @@
+  private final Set<Extract> createdInstances;
+  private final DateTimeFormatter dtf;
+
+  public ExtractFactory(String dateTimeFormat) {
```
Please add some unit tests for this new class.

chavdar commented on gobblin-api/src/main/java/gobblin/configuration/SourceState.java:
```diff
@@ -49,7 +49,7 @@
 public class SourceState extends State {
 
-  private static final Set<Extract> extractSet = Sets.newConcurrentHashSet();
+  private static final Set<Extract> EXTRACT_SET = Sets.newConcurrentHashSet();
```
Why don't make use ExtractFactory?

zliu41 replied:
`SourceState.createExtract()` invokes a deprecated constructor of `Extract` which adds properties in `SourceState` to the created `Extract`. `ExtractFactory` doesn't do this. Changing this may break existing jobs.

zliu41 commented:
Updated

chavdar commented:
LGTM.
